examples/cgroup-rstat-flushing.bpf.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Measure cgroup rstat (recursive stats) flushing overhead and latency.
+ *
+ * Loosely based on bpftrace script cgroup_rstat_tracepoint.bt
+ *  - https://github.com/xdp-project/xdp-project/blob/master/areas/latency/cgroup_rstat_tracepoint.bt
+ *
+ * Depends on tracepoints added in kernel v6.10
+ */
+
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "maps.bpf.h"
+
+#define MAX_CGRP_LEVELS	5
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, MAX_CGRP_LEVELS + 1);
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_flush_total SEC(".maps");
+
+/* Total counter for obtaining lock together with contended state
+ *
+ * This counter also contains "yield" case. To determine "normal" lock
+ * case subtract "yield" counter in prometheus query.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 2); /* contended state used as key */
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_locked_total SEC(".maps");
+
+/* Counter for obtaining lock again after yield (and contended state).
+ *
+ * Kernel can yield the rstat lock when walking individial CPU stats.
+ * This leads to "interresting" concurrency issues.  Thus, keep
+ * seperate counter for "yield" cases to help diagnose.
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 2); /* contended state used as key */
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_locked_yield SEC(".maps");
+
+/* Counter for lock contended case, recorded per cgroup level */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, MAX_CGRP_LEVELS + 1);
+	__type(key, u32);
+	__type(value, u64);
+} cgroup_rstat_lock_contended SEC(".maps");
+
+/** Measurement#1: lock rates
+ *  =========================
+ For locks, the problematic/interesting case is when the lock was contended.
+
+ Simply counting the lock+unlock is complicated by yielding the lock in the main
+ flushing loop (in cgroup_rstat_flush_locked()). The tracepoints "cpu" argument
+ will be (minus) -1 when the lock is not a yielded lock.
+
+ Concern: The lock rates will vary (a lot) and aggregating this as an average
+ will not capture spikes. Especially given Prometheus capture intervals only
+ happens every 53 seconds.
+
+ Q: will lock "type=yield" be a good idea?
+
+*/
+
+SEC("tp_btf/cgroup_rstat_locked")
+int BPF_PROG(rstat_locked, struct cgroup *cgrp, int cpu, bool contended)
+{
+	u32 key = contended;
+	u64 *cnt;
+
+	read_array_ptr(&cgroup_rstat_locked_total, &key, cnt);
+	(*cnt)++;
+
+	if (cpu >= 0) {
+		read_array_ptr(&cgroup_rstat_locked_yield, &key, cnt);
+		(*cnt)++;
+	}
+
+	/* What cgrp level is interesting, but I didn't manage to encode it in
```

Comment:
nit: `cgrp` -> `cgroup`

Reply:
Sure to nit.
I was hoping to get your input to the contents of the comment:
```
	/* What cgrp level is interesting, but I didn't manage to encode it in
	 * above counters.  As contended case is the most interesting, have
	 * level counter for contended.
	 */
```
When coding BPF, I usually do **one** map lookup to get a stats `struct`, that I will update with multiple stats.
Here I end-up doing 3 map lookups for single counters.

When I stated coding I though I could represent something like:
```
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="0"} 6558
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="0"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="1"} 34
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="1"} 123
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0" level="5"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1" level="5"} 0
```
Question: But ...
1. I cannot figure out how to get ebpf_exporter to generate this?
2. I don't know if these counters makes sense for Prometheus ?

Reply:
> I don't know if these counters makes sense for Prometheus ?

They do

> I cannot figure out how to get ebpf_exporter to generate this?

Each line here is a separate key in a single map called `ebpf_exporter_cgroup_rstat_locked_total`:

```
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="0"} 6558
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="0"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="1"} 34
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="1"} 123
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="2"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="3"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="4"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="0", level="5"} 0
ebpf_exporter_cgroup_rstat_locked_total{contended="1", level="5"} 0
```

The mapping is effectively:

* `{contended="0", level="0"}` -> `6558`
* `{contended="1", level="0"}` -> `0`
* `{contended="0", level="1"}` -> `34`
* ...

To encode each key, you turn the raw bytes of the key into the values with the help of decoders.

Working backwards from what you expect on the output, you want something like this in the config:

```yaml
    - name: cgroup_rstat_locked_total
      help: Total number of times rstat lock was obtainted and contended state
      labels:
        - name: contended
          size: 1
          decoders:
            - name: uint
        - name: level
           size: 1
           decoders:
            - name: uint
```

On the kernel side you'd want something like this to represent the key:

```c
struct key_t {
    u8 contended;
    u8 level;
};
```

The names here do not matter, only sizes do (u8 corresponds to `size: 1` in the config).

Given that your keys are complex structs and not small numbers, you cannot store these in an array, so you need a hashmap:

```
struct {
    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
    __uint(max_entries, 128);
    __type(key, struct key_t);
    __type(value, u64);
} cgroup_rstat_locked_total SEC(".maps");
```

To actually increment a value, you need to construct the key and to the increment in the map:

```c
// construct however you need to
struct key_t key = {};
key.contended = 1;
key.level = 2;

increment_map_nosync(&cgroup_rstat_locked_total, &key, 1);
```

Here's an example with a complex key:

* https://github.com/cloudflare/ebpf_exporter/blob/master/examples/pci.yaml
* https://github.com/cloudflare/ebpf_exporter/blob/master/examples/pci.bpf.c

Hope this makes sense.

Reply:
These suggestions worked for encoding more information ("labels" in Prometheus terms). 
I've coded it up and pushed as part of this patchset/PR.
(p.s. I've also encoded "yield" state)